File: sdk/ts/src/upgrade.ts
/**
 * @packageDocumentation
 *
 * Version check and notify — compares the running tool version against the
 * latest GitHub release and writes a human-readable upgrade notice to any
 * {@link https://nodejs.org/api/stream.html#class-streamwritable | Writable}
 * stream when a newer release is available.
 *
 * ### Cache behaviour
 * Results are persisted to `{stateDir}/.upgrade-{name}-cache.json`.
 * The cache stores `{ version, checkedAt }`. On the next call within `cacheTTL`
 * milliseconds the remote API is not contacted; the cached version is used
 * directly. After the TTL expires the cache is refreshed on the next call.
 * Cache I/O errors are swallowed — a stale or missing cache simply causes a
 * fresh fetch.
 *
 * ### Error handling philosophy
 * Version checking is a best-effort, non-critical operation.  Any failure —
 * network error, bad JSON, HTTP error, filesystem error, invalid semver — is
 * silently discarded.  The function always resolves without throwing.  If
 * something goes wrong the user sees neither a notice nor an error.
 *
 * ### Notice format
 * ```
 * \nA new release of {name} is available: v{current} → v{latest}\nhttps://github.com/{owner}/{repo}/releases/latest\n
 * ```
 *
 * ### Usage example
 * ```ts
 * import { createChecker } from '@hop-top/kit/upgrade';
 *
 * const checker = createChecker({
 *   name: 'my-cli',
 *   currentVersion: '1.2.3',
 *   owner: 'myorg',
 *   repo: 'my-cli',
 * });
 *
 * // At the end of your CLI command:
 * await checker.notifyIfAvailable(process.stderr);
 * ```
 */

import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import semver from 'semver';

// ─── Public types ─────────────────────────────────────────────────────────────

/**
 * Static configuration for a version {@link Checker}.
 */
export interface CheckerOptions {
  /** Human-readable tool name shown in the upgrade notice. */
  name: string;
  /** Currently running version string (semver, without leading `v`). */
  currentVersion: string;
  /** GitHub organisation or username that owns the repository. */
  owner: string;
  /** GitHub repository name. */
  repo: string;
  /**
   * How long (ms) a cached result stays valid before a fresh fetch is made.
   * @defaultValue 86_400_000 (24 hours)
   */
  cacheTTL?: number;
  /**
   * Directory used to store the cache file.
   * @defaultValue `os.tmpdir()`
   */
  stateDir?: string;
  /**
   * HTTP request timeout in milliseconds.
   * @defaultValue 10_000
   */
  timeout?: number;
}

/**
 * Returned by {@link createChecker}.  Call {@link Checker.notifyIfAvailable}
 * at the end of a CLI command to surface upgrade notices non-intrusively.
 */
export interface Checker {
  /**
   * Checks for a newer GitHub release and writes a one-line notice to `out`
   * if one is available.  Silently swallows all errors.
   *
   * @param out - Any Node.js writable stream (e.g. `process.stderr`).
   */
  notifyIfAvailable(out: NodeJS.WritableStream): Promise<void>;
}

// ─── Internal types ───────────────────────────────────────────────────────────

interface CacheEntry {
  version: string;
  checkedAt: number;
}

// ─── Implementation ───────────────────────────────────────────────────────────

/**
 * Creates a {@link Checker} configured with the provided options.
 *
 * The checker is stateless beyond what is persisted in the cache file; it is
 * safe to create multiple instances with the same options.
 *
 * @param opts - Configuration options; see {@link CheckerOptions}.
 * @returns A {@link Checker} ready to call.
 */
export function createChecker(opts: CheckerOptions): Checker {
  const {
    name,
    currentVersion,
    owner,
    repo,
    cacheTTL = 24 * 60 * 60 * 1000,
    stateDir = os.tmpdir(),
    timeout = 10_000,
  } = opts;

  const cacheFile = path.join(stateDir, `.upgrade-${name}-cache.json`);
  const releaseURL = `https://api.github.com/repos/${owner}/${repo}/releases/latest`;
  const releasesPageURL = `https://github.com/${owner}/${repo}/releases/latest`;

  // ── cache helpers ──────────────────────────────────────────────────────────

  function readCache(): CacheEntry | null {
    try {
      const raw = fs.readFileSync(cacheFile, 'utf8');
      const parsed = JSON.parse(raw) as CacheEntry;
      if (typeof parsed.version === 'string' && typeof parsed.checkedAt === 'number') {
        return parsed;
      }
    } catch {
      // Missing or malformed cache — treat as miss.
    }
    return null;
  }

  function writeCache(version: string): void {
    try {
      const entry: CacheEntry = { version, checkedAt: Date.now() };
      fs.writeFileSync(cacheFile, JSON.stringify(entry), 'utf8');
    } catch {
      // Cache write failure is non-fatal.
    }
  }

  // ── fetch latest version ───────────────────────────────────────────────────

  async function fetchLatestVersion(): Promise<string | null> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeout);
    try {
      const res = await fetch(releaseURL, {
        signal: controller.signal,
        headers: { Accept: 'application/vnd.github+json' },
      });
      if (!res.ok) return null;
      const body = (await res.json()) as { tag_name?: string };
      const tag = body.tag_name;
      if (typeof tag !== 'string') return null;
      // Strip leading 'v' so semver.gt works with or without prefix.
      return tag.replace(/^v/, '');
    } finally {
      clearTimeout(timer);
    }
  }

  // ── notifyIfAvailable ──────────────────────────────────────────────────────

  async function notifyIfAvailable(out: NodeJS.WritableStream): Promise<void> {
    try {
      let latestVersion: string | null = null;

      // Check cache first.
      const cached = readCache();
      if (cached !== null && Date.now() - cached.checkedAt < cacheTTL) {
        latestVersion = cached.version;
      } else {
        latestVersion = await fetchLatestVersion();
        if (latestVersion !== null) {
          writeCache(latestVersion);
        }
      }

      if (latestVersion === null) return;

      // Only notify when latest is strictly greater than current.
      if (!semver.gt(latestVersion, currentVersion)) return;

      const notice =
        `\nA new release of ${name} is available: v${currentVersion} → v${latestVersion}\n` +
        `${releasesPageURL}\n`;

      out.write(notice);
    } catch {
      // Any unexpected error is silently swallowed.
    }
  }

  return { notifyIfAvailable };
}

File: sdk/py/hop_top_kit/upgrade.py
"""
hop_top_kit.upgrade — version check and notify for hop-top Python tools.

Fetches the latest GitHub release for a tool and writes a human-readable
upgrade notice to an output stream when a newer version is available.

Cache behaviour:
    Results are cached in a JSON file at
    ``{state_dir}/.upgrade-{name}-cache.json`` with the fields
    ``version`` (str) and ``checked_at`` (Unix timestamp float).
    While the cached entry is younger than ``cache_ttl``, no network
    request is made.  When the cache is absent or expired the GitHub
    Releases API is queried; a fresh entry is written on success.

Error handling:
    All errors (network, JSON parse, file I/O, version parse) are
    silently swallowed.  The function is intended for best-effort use
    at CLI startup and must never abort the tool.

Notice format (written to ``out`` when a newer version exists)::

    \\nA new release of {name} is available: {current} → {latest}\\n
    https://github.com/{owner}/{repo}/releases/latest\\n

Usage example::

    from datetime import timedelta
    from hop_top_kit.upgrade import CheckerOptions, create_checker

    opts = CheckerOptions(
        name="mytool",
        current_version="1.2.3",
        owner="myorg",
        repo="mytool",
        cache_ttl=timedelta(hours=24),
    )
    checker = create_checker(opts)

    # At CLI startup — prints notice if newer version exists, else silent:
    checker.notify_if_available()
"""

from __future__ import annotations

import contextlib
import json
import os
import sys
import tempfile
import time
import urllib.request
from dataclasses import dataclass, field
from datetime import timedelta
from typing import IO

from packaging.version import InvalidVersion, Version


@dataclass
class CheckerOptions:
    """Static configuration for an upgrade Checker.

    Attributes:
        name:            Package / tool name displayed in the notice.
        current_version: Installed version string (e.g. ``"1.2.3"``).
        owner:           GitHub repository owner (org or user).
        repo:            GitHub repository name.
        cache_ttl:       How long to reuse a cached check result before
                         hitting the network again.  Default: 24 hours.
        state_dir:       Directory used for the cache file.  Defaults to
                         ``tempfile.gettempdir()`` when ``None``.
        timeout:         HTTP request timeout in seconds.  Default: 5.
    """

    name: str
    current_version: str
    owner: str
    repo: str
    cache_ttl: timedelta = field(default_factory=lambda: timedelta(hours=24))
    state_dir: str | None = None
    timeout: int = 5


class Checker:
    """Checks whether a newer release is available and notifies the user.

    Constructed via :func:`create_checker`; do not instantiate directly.
    """

    def __init__(self, opts: CheckerOptions) -> None:
        self._opts = opts

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def notify_if_available(self, out: IO = sys.stdout) -> None:
        """Write an upgrade notice to *out* if a newer version is available.

        Silently no-ops on any error (network, I/O, parse, etc.).
        Uses a local cache to avoid redundant network requests.

        Args:
            out: Output stream to write the notice to.
                 Defaults to ``sys.stdout``.
        """
        with contextlib.suppress(Exception):
            self._run(out)

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    def _cache_path(self) -> str:
        state_dir = self._opts.state_dir or tempfile.gettempdir()
        return os.path.join(state_dir, f".upgrade-{self._opts.name}-cache.json")

    def _read_cache(self) -> str | None:
        """Return cached version string if still within TTL, else None."""
        path = self._cache_path()
        try:
            with open(path) as f:
                data = json.load(f)
            checked_at: float = data["checked_at"]
            ttl_seconds = self._opts.cache_ttl.total_seconds()
            if time.time() - checked_at < ttl_seconds:
                return str(data["version"])
        except Exception:
            pass
        return None

    def _write_cache(self, version: str) -> None:
        """Persist *version* and current timestamp to the cache file."""
        path = self._cache_path()
        try:
            with open(path, "w") as f:
                json.dump({"version": version, "checked_at": time.time()}, f)
        except Exception:
            pass

    def _fetch_latest(self) -> str | None:
        """Fetch the latest release tag from GitHub and return version string.

        Returns ``None`` on any error or non-200 status.
        """
        url = f"https://api.github.com/repos/{self._opts.owner}/{self._opts.repo}/releases/latest"
        with urllib.request.urlopen(url, timeout=self._opts.timeout) as resp:
            if resp.status != 200:
                return None
            body = resp.read()
        data = json.loads(body)
        tag: str = data["tag_name"]
        return tag.lstrip("v")

    def _run(self, out: IO) -> None:
        """Core logic: fetch (or read cache), compare, maybe write notice."""
        opts = self._opts

        latest = self._read_cache()
        if latest is None:
            latest = self._fetch_latest()
            if latest is None:
                return
            self._write_cache(latest)

        try:
            current_v = Version(opts.current_version)
            latest_v = Version(latest)
        except InvalidVersion:
            return

        if latest_v <= current_v:
            return

        out.write(
            f"\nA new release of {opts.name} is available: "
            f"{opts.current_version} \u2192 {latest}\n"
            f"https://github.com/{opts.owner}/{opts.repo}/releases/latest\n"
        )


def create_checker(opts: CheckerOptions) -> Checker:
    """Create a :class:`Checker` from *opts*.

    Args:
        opts: Configuration for the checker.

    Returns:
        A :class:`Checker` instance ready to call
        :meth:`~Checker.notify_if_available`.

    Example::

        from datetime import timedelta
        from hop_top_kit.upgrade import CheckerOptions, create_checker

        checker = create_checker(CheckerOptions(
            name="mytool",
            current_version="0.9.0",
            owner="myorg",
            repo="mytool",
        ))
        checker.notify_if_available()
    """
    return Checker(opts)
